integrations/tensorflow/bindings/python/pyiree/tf/support/tf_utils.py

```diff
@@ -259,7 +280,7 @@ def __init__(self,
                backend_info: "BackendInfo",
                exported_names: Sequence[str] = (),
                artifacts_dir: str = None,
-               _create_reinitialized_args: Sequence[Any] = None):
+               _create_reinitialized_dict: Dict[str, Any] = None):
```

Comment:
@stellaraccident is there a better Python pattern for this kind of internal constructor than "secret squirrel argument to the constructor"?

Reply:
🐿️

My suggestion on https://github.com/google/iree/pull/2897 is to just do away with this in favor of storing the reusable stateless parts of this in separate class. This also seems to be a violation of the "constructors should do no work" principle, which I think is relatively cross-language.

What about:

CompiledModule: contains static module blobs and such. Has a module_blob backend_info
RunnableModule: the thing you can actually call stuff on. Has a module, context, config, etc.

The RunnableModule constructor could do the conversion, or there could be a separate function to do it and the runnable module constructor could just be KISS `self.foo = foo`...

Reply:
I'm not a huge fan of requiring someone to create a CompilationInfo object and manually pass it to a seperate class in order to get something that they can actually use. If I were using IREE at the python level, I would expect to be able to do something like this:

```python3
module = IreeCompiledModule(MyModuleClass, BackendInfo['iree_vmla'])
module.predict(...)
```

I am pretty annoyed by the lack of multiple constructors in python though, and really that's all this squirrel is emulating. It makes sense given their lax approach to typing, but I would love a better alternative.

Reply:
You can emulate multiple constructors with static methods though. It's actually an even stronger argument for simple constructors IMO.

```python
module = IreeCompiledModule.create(MyModuleClass, BackendInfo['iree_vmla'])
```

Reply:
That's what we had [here](https://github.com/google/iree/commit/7a5b8c969c8159f826119e95fbecfa1fbdc92b03#diff-742129371bdef91e335f7e42ee60b577L139), I just found using static compilers in python a bit unintuitive and verbose. There wasn't a way to signal to the user that they were incorrectly using the class if they call it's default constructor (because the static ones need to use it too). I sort of got around this by making the default constructor unpleasant to use (no default args, even for the non-required ones), but it's still reflexive to assume the answer to "how do I make an instance of this" in python is just "call it's constructor". That's true even when it probably shouldn't be (e.g. `tf.keras.Model` violates the _hell_ out of the "constructors don't do work" principle). IDK, I'm definitely overthinking this. 😅

Reply:
This is one of those situations where someone with Python "readability" would be nice.

I'm not opposed to what we have here, just was hoping maybe Stella could chime in with a simple suggestion for how to approach this API wise.

Reply:
Deduping my comment from the other PR:

> Seems like it's easy to provide a function for compiling and then initializing if we have those things separate, but hard to break things out if you don't. I would rather the init be simple and do no initialization. One option would be to still have one class, but have a separate initialize function that has to be called before it's used and then a createInitialized function that does both. initialize functions are usually a bad API though because then anytime you do anything with it you have to make sure it's initialized first. Hence just having a separate class for it.

Reply:
100% agree on `initialize` methods that aren't `__init__` or automatically called by `__init__` being bad. Everything related to compilation is currently separated out into `compile_tf_module`, and `IreeCompiledModule` just wraps that call instead of taking it's output for simplicity.

We could do away with one of these squirrels by removing the option to create a new `IreeCompiledModule` from an old one, and just resetting the context in an `reinitialize` method. That doesn't address the one in `tf_test_utils.Trace.load` however.

Reply:
Yeah I'm skeptical of that "simplicity" :-P

Something like this seems pretty simple to me, but I agree with Sean that it would be good to have a "pythonista" weigh in :stuck_out_tongue: 

```python
class CompiledModule(object):
  def __init__(self, module_name, backend):
    self.module_name = module_name
    self.backend = backend

  @classmethod
  def create_compiled(cls, module_class, backend_info, exported_names, artifacts_dir):
    raise NotImplementedError()

class IreeCompiledModule(CompiledModule):
  def __init(self, module_name, backend_info, module_blob):
    super().__init__(module_name, backend_info.name)
    self._module = rt.VmModule.from_flatbuffer(module_blob)
    self._config = rt.Config(driver_name=backend_info.driver)
    self._context = rt.SystemContext(
        modules=[self._module], config=self._config)

  @classmethod
  def create_compiled(module_class, backend_info, exported_names, artifacts_dir):
    module_blob = compile_tf_module(
          tf_module=module_class(),
          backend_infos=[backend_info],
          exported_names=exported_names,
          artifacts_dir=artifacts_dir)
    return IreeCompiledModule(module_class.__name__, backend_info, exported_names, module_blob)

class TfCompiledModule(CompiledModule):
  def __init__(self, module_name, backend_info, tf_module):
    super().__init__(module_name, backend_info.name)
    self._tf_module = tf_module

  @classmethod
  def create_compiled(cls, module_class, backend_info, exported_names, artifacts_dir):
    return TfCompiledModule(module_class.__name__, backend_info, module_class())
```

Reply:
Maybe we should open a separate issue for this discussion? I'd still appreciate @stellaraccident weighing in.